core/src/local/persistent_state.rs

```diff
@@ -0,0 +1,891 @@
+        }
+
+        // IndexID is a u32, so bincode uses 4 bytes to serialize it (which we'll skip past):
+        let start = 4;
+        // We encoded the size of the key itself with a u64, which bincode uses 8 bytes to encode:
+        let size_offset = start + 8;
+        let key_size: u64 = bincode::deserialize(&key[start..size_offset]).unwrap();
+        let prefix_len = size_offset + key_size as usize;
+        let mut bytes = Vec::from(key);
+
+        // Strip away the IndexEpoch and IndexSeq if we haven't already done so:
+        if key.len() > prefix_len {
+            bytes.truncate(prefix_len);
+        }
+
+        bytes
+    }
+
+    // Decides which keys the prefix transform should apply to.
+    fn in_domain_fn(key: &[u8]) -> bool {
+        key != META_KEY
+    }
+
+    // A key is built up of five components:
+    //  * `index_id`
+    //     Uniquely identifies the index in self.indices.
+    //  * `row_size`
+    //      The byte size of `row` when serialized with binode.
+    //  * `row`
+    //      The actual index values.
+    //  * `epoch`
+    //      Incremented on each PersistentState initialization.
+    //  * `seq`
+    //      Sequence number since last epoch.
+    fn serialize_key(&self, index_id: IndexID, row: &Vec<&DataType>, seq: IndexSeq) -> Vec<u8> {
+        let size: u64 = bincode::serialized_size(&row).unwrap();
+        bincode::serialize(&(index_id, size, row, self.epoch, seq)).unwrap()
+    }
+
+    // Used with DB::prefix_iterator to go through all the rows for a given key.
+    fn serialize_prefix(index: IndexID, key: &Vec<&DataType>) -> Vec<u8> {
+        let size: u64 = bincode::serialized_size(&key).unwrap();
+        bincode::serialize(&(index, size, key)).unwrap()
+    }
+
+    // Filters out secondary indices to return an iterator for the actual key-value pairs.
+    fn all_rows(&self) -> impl Iterator<Item = (Box<[u8]>, Box<[u8]>)> {
+        self.db
+            .as_ref()
+            .unwrap()
+            .full_iterator(rocksdb::IteratorMode::Start)
+            .filter(|&(ref key, _)| {
+                // Filter out non-pk indices:
+                let i: IndexID = bincode::deserialize(&key).unwrap();
+                i == 0
+            })
+    }
+
+    // Puts by primary key first, then retrieves the existing value for each index and appends the
+    // newly created primary key value.
+    // TODO(ekmartin): This will put exactly the values that are given, and can only be retrieved
+    // with exactly those values. I think the regular state implementation supports inserting
+    // something like an Int and retrieving with a BigInt.
+    fn insert(&mut self, batch: &mut WriteBatch, r: Vec<DataType>) {
+        for index in self.indices.iter_mut() {
+            index.seq += 1;
+        }
+
+        let (secondary_indices, serialized_pk) = if let Some(ref pk_cols) = self.primary_key {
+            let pk = pk_cols.iter().map(|i| &r[*i]).collect::<Vec<_>>();
```

Comment:
I *think* that if `serialize_prefix` took a `KeyType`, you wouldn't need to `collect()` (and allocate) here.

Reply:
Yeah, good idea. Changed `serialize_key` and `serialize_prefix` to take a `KeyType` - definitely nicer.